internal/component/prometheus/remote/queue/serialization/serializer.go

```diff
@@ -0,0 +1,223 @@
+	"github.com/vladopajic/go-actor/actor"
+)
+
+// serializer collects data from multiple appenders and will write them to file.Storage.
+// serializer will trigger based on the last flush duration OR if it hits a certain amount of items.
+type serializer struct {
+	inbox               actor.Mailbox[*types.TimeSeriesBinary]
+	metaInbox           actor.Mailbox[*types.TimeSeriesBinary]
+	cfgInbox            actor.Mailbox[types.SerializerConfig]
+	maxItemsBeforeFlush int
+	flushFrequency      time.Duration
+	queue               types.FileStorage
+	lastFlush           time.Time
+	logger              log.Logger
+	self                actor.Actor
+	// Every 1 second we should check if we need to flush.
+	flushTestTimer *time.Ticker
+	series         []*types.TimeSeriesBinary
+	meta           []*types.TimeSeriesBinary
+	msgpBuffer     []byte
+	stats          func(stats types.SerializerStats)
+}
+
+func NewSerializer(cfg types.SerializerConfig, q types.FileStorage, stats func(stats types.SerializerStats), l log.Logger) (types.Serializer, error) {
+	s := &serializer{
+		maxItemsBeforeFlush: int(cfg.MaxSignalsInBatch),
+		flushFrequency:      cfg.FlushFrequency,
+		queue:               q,
+		series:              make([]*types.TimeSeriesBinary, 0),
+		logger:              l,
+		inbox:               actor.NewMailbox[*types.TimeSeriesBinary](),
+		metaInbox:           actor.NewMailbox[*types.TimeSeriesBinary](),
+		cfgInbox:            actor.NewMailbox[types.SerializerConfig](),
+		flushTestTimer:      time.NewTicker(1 * time.Second),
+		msgpBuffer:          make([]byte, 0),
+		lastFlush:           time.Now(),
+		stats:               stats,
+	}
+
+	return s, nil
+}
+func (s *serializer) Start() {
+	// All the actors and mailboxes need to start.
+	s.queue.Start()
+	s.self = actor.Combine(actor.New(s), s.inbox, s.metaInbox, s.cfgInbox).Build()
+	s.self.Start()
+}
+
+func (s *serializer) Stop() {
+	s.queue.Stop()
+	s.self.Stop()
+}
+
+func (s *serializer) SendSeries(ctx context.Context, data *types.TimeSeriesBinary) error {
+	return s.inbox.Send(ctx, data)
+}
+
+func (s *serializer) SendMetadata(ctx context.Context, data *types.TimeSeriesBinary) error {
+	return s.metaInbox.Send(ctx, data)
+}
+
+func (s *serializer) UpdateConfig(ctx context.Context, cfg types.SerializerConfig) error {
+	return s.cfgInbox.Send(ctx, cfg)
+}
+
+func (s *serializer) DoWork(ctx actor.Context) actor.WorkerStatus {
+	// Check for config which should have priority. Selector is random but since incoming
+	// series will always have a queue by explicitly checking the config here we always give it a chance.
+	// By pulling the config from the mailbox we ensure it does NOT need a mutex around access.
+	select {
+	case <-ctx.Done():
+		return actor.WorkerEnd
+	case cfg, ok := <-s.cfgInbox.ReceiveC():
+		if !ok {
+			return actor.WorkerEnd
+		}
+		s.maxItemsBeforeFlush = int(cfg.MaxSignalsInBatch)
+		s.flushFrequency = cfg.FlushFrequency
```

Comment:
Could we simplify and use atomic variables instead of the `cfgInbox` etc? Or a single atomic variable to hold the SerializerConfig. Would make the code a bit simpler and probably similar performance.

Reply:
In this case we could since we don't generally take action on the config, though this mirrors what we do in other objects so I am inclined to leave it like this for consistency.